src/pages/permissions/panels/EditIdentityGroupsPanel.tsx

```diff
@@ -0,0 +1,416 @@
+import {
+  ActionButton,
+  Button,
+  Icon,
+  useNotify,
+} from "@canonical/react-components";
+import { useQuery, useQueryClient } from "@tanstack/react-query";
+import { fetchLxdGroups } from "api/auth-groups";
+import SidePanel from "components/SidePanel";
+import { FC, useEffect, useState } from "react";
+import { queryKeys } from "util/queryKeys";
+import usePanelParams from "util/usePanelParams";
+import ScrollableTable from "components/ScrollableTable";
+import SelectableMainTable from "components/SelectableMainTable";
+import {
+  generateGroupAllocationsForIdentities,
+  getCurrentGroupAllocationsForIdentities,
+} from "util/permissionIdentities";
+import useSortTableData from "util/useSortTableData";
+import { useToastNotification } from "context/toastNotificationProvider";
+import useEditHistory from "util/useEditHistory";
+import IdentityGroupsPanelActions from "./IdentityGroupsPanelActions";
+import { pluralize } from "util/instanceBulkActions";
+import IdentityGroupsPanelConfirmModal from "./IdentityGroupsPanelConfirmModal";
+import { updateGroupsForIdentities } from "api/auth-identities";
+import { LxdIdentity } from "types/permissions";
+import { MainTableRow } from "@canonical/react-components/dist/components/MainTable/MainTable";
+import PermissionGroupsFilter from "../PermissionGroupsFilter";
+
+type PanelHistoryState = {
+  groupsAllocatedToAll: string[];
+  groupsAllocatedToNone: string[];
+};
```

Comment:
```suggestion
type GroupEditHistory = {
  groupsAdded: string[];
  groupsRemoved: string[];
};
```
I think there was a misunderstanding here. I think the complexity can be reduced by keeping the user requested changes in the state. Then the logic below to calculate modified groups and the current status of each group checkbox should become much more straight forward. Wdyt?

Reply:
Thanks! Got a wave of inspiration, this definitely simplifies things. I've got it working now so just need to cleanup the code a bit.